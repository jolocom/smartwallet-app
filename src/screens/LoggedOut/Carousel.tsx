import React, { useState } from 'react'
import {
  View,
  StyleSheet,
  Dimensions,
  ImageBackground,
  Platform,
} from 'react-native'
import Carousel, { Pagination } from 'react-native-snap-carousel'
import { Walkthrough1, Walkthrough2, Walkthrough3 } from '~/assets/images'
import useTranslation from '~/hooks/useTranslation'
import AbsoluteBottom from '~/components/AbsoluteBottom'
import JoloText, { JoloTextKind, JoloTextWeight } from '~/components/JoloText'
import { JoloTextSizes } from '~/utils/fonts'
import { Colors } from '~/utils/colors'
import BP from '~/utils/breakpoints'

const SLIDER_WIDTH = Dimensions.get('window').width + 80
const ITEM_WIDTH = Dimensions.get('window').width

const CustomCarousel = () => {
  const { t } = useTranslation()
  const isCarousel = React.useRef(null)

  const [index, setIndex] = useState(0)

  const walkthroughData = [
    {
      background: Walkthrough1,
      header: t('Walkthrough.titleOne'),
      paragraph: t('Walkthrough.descriptionOne'),
    },
    {
      background: Walkthrough2,
      header: t('Walkthrough.titleTwo'),
      paragraph: t('Walkthrough.descriptionTwo'),
    },
    {
      background: Walkthrough3,
      header: t('Walkthrough.titleThree'),
      paragraph: t('Walkthrough.descriptionThree'),
    },
  ]

  const CarouselCardItem = ({ item, index }) => {
    return (
      <React.Fragment>
        <ImageBackground
          key={index}
          style={styles.background}
          source={item.background}
        />
        <AbsoluteBottom
          customStyles={{
            ...styles.consistentContainer,
            bottom: Platform.select({
              ios: BP({ default: 165, medium: 180, large: 180 }),
              android: 235,
            }),
          }}
        >
          <JoloText
            kind={JoloTextKind.title}
            size={JoloTextSizes.big}
            weight={JoloTextWeight.regular}
            color={Colors.white90}
          >
            {item.header}
          </JoloText>
          <JoloText
            kind={JoloTextKind.subtitle}
            size={JoloTextSizes.big}
            color={Colors.white80}
            customStyles={{ marginTop: 12 }}
          >
            {item.paragraph}
          </JoloText>
        </AbsoluteBottom>
      </React.Fragment>
    )
  }

  return (
    <View>
      <Carousel
        ref={isCarousel}
        data={walkthroughData}
        renderItem={CarouselCardItem}
        sliderWidth={SLIDER_WIDTH}
        itemWidth={ITEM_WIDTH}
        autoplay={true}
        autoplayDelay={1000}
        autoplayInterval={8000}
        enableMomentum={false}
        lockScrollWhileSnapping={true}
        loop={true}
        onSnapToItem={(index) => setIndex(index)}
        useScrollView={true}
      />
      <View style={styles.dotsContainer}>
        <Pagination
          dotsLength={walkthroughData.length}
          activeDotIndex={index}
          carouselRef={isCarousel}
          dotStyle={styles.activeDotStyle}
          inactiveDotStyle={styles.inactiveDotStyle}
          inactiveDotOpacity={0.4}
          inactiveDotScale={0.6}
          tappableDots={true}
          animatedDuration={0}
        />
      </View>
    </View>
  )
}

export default CustomCarousel

const styles = StyleSheet.create({
  background: { ...(StyleSheet.absoluteFill as {}), top: -20 },
  consistentContainer: {
    paddingHorizontal: '5%',
  },
  dotsContainer: {
    position: 'relative',
    ...Platform.select({
      android: {
        bottom: 175,
      },
      ios: {
        bottom: 200,
      },
    }),
  },
  activeDotStyle: {
    width: 6,
    height: 6,
    borderRadius: 3,
    marginHorizontal: 0,
    backgroundColor: Colors.floralWhite,
  },
  inactiveDotStyle: { backgroundColor: Colors.peach },
})
