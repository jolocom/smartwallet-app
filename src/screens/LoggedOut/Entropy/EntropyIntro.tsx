import React from 'react'
import { View, StyleSheet } from 'react-native'

import { HandAnimation } from './HandAnimation'
import Header, { HeaderSizes } from '~/components/Header'
import Paragraph, { ParagraphSizes } from '~/components/Paragraph'
import { strings } from '~/translations/strings'
import { Colors } from '~/utils/colors'

export const EntropyIntro: React.FC = () => {
  return (
    <View style={styles.container}>
      <View style={styles.handContainer}>
        <HandAnimation />
      </View>
      <View style={styles.textContainer}>
        <Header color={Colors.white85}>{strings.SET_UP_YOUR_IDENTITY}</Header>
        <Paragraph
          size={ParagraphSizes.medium}
          color={Colors.white70}
          customStyles={styles.paragraph}
        >
          {strings.TAP_THE_SCREEN_AND_DRAW_RANDOMLY_ON_IT_UNTIL_YOU_COLLECT_100}
        </Paragraph>
      </View>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  handContainer: {
    margin: 60,
    width: 78,
    height: 84,
  },
  textContainer: {
    alignItems: 'center',
  },
  paragraph: {
    opacity: 0.8,
    paddingHorizontal: 20,
  },
})
