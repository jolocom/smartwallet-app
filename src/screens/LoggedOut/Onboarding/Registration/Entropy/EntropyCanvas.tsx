import React, { useState, useRef, useEffect } from 'react'
import { PanResponder, GestureResponderEvent } from 'react-native'
import { Svg, Path, Rect, Circle } from 'react-native-svg'

import { useForceUpdate } from '~/hooks/generic'
import {
  findIntersections,
  Coordinates,
  extractCoords,
  shouldComputeLine,
} from './utils'
import { Colors } from '~/utils/colors'

interface Props {
  disabled: boolean
  addPoint: (x: number, y: number) => void
}

export const MIN_DISTANCE_SQ = 30
export const MAX_LINE_PTS = 100

export const EntropyCanvas: React.FC<Props> = React.memo(
  ({ disabled, addPoint }) => {
    const {
      handleDraw,
      handleDrawStart,
      pathDs,
      pathEls,
      circles,
    } = useCanvasGestures(addPoint)

    const panResponder = PanResponder.create({
      onMoveShouldSetPanResponder: () => true,
      onMoveShouldSetPanResponderCapture: () => true,
      onPanResponderGrant: handleDrawStart,
      onPanResponderMove: handleDraw,
    })

    return (
      <>
        <Svg
          width="100%"
          height="100%"
          {...(!disabled && panResponder.panHandlers)}
        >
          <Rect width="100%" height="100%" opacity="0.1"></Rect>
          {pathDs.current.map((d, idx) => {
            if (!d) return null
            return (
              <Path
                key={idx}
                ref={(el) => (pathEls[idx] = el)}
                d={d}
                fill="none"
                stroke={Colors.white}
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeDasharray="1,10"
                strokeWidth="2"
              />
            )
          })}
          {circles.map(([x, y], i) => (
            <Circle cx={x} cy={y} key={i} r={4} fill={Colors.white} />
          ))}
        </Svg>
      </>
    )
  },
)

export const useCanvasGestures = (action: (x: number, y: number) => void) => {
  const forceUpdate = useForceUpdate()

  const [circles, setCircles] = useState<number[][]>([])
  const [, setCirclesN] = useState<number>(0)

  const pathEls = useRef<any[]>(new Array(10)).current
  const pathDs = useRef<string[]>([])
  const pathIdx = useRef<number>(0)
  const linesPts = useRef<number[]>([])
  const linesPtsIdx = useRef<number>(0)
  const coords = useRef<Coordinates>({
    prevX: 0,
    prevY: 0,
    curX: 0,
    curY: 0,
  })

  useEffect(() => {
    for (let i = 0; i < pathEls.length; i++) {
      pathDs.current = [...pathDs.current, '']
    }
  }, [])

  const modifyPath = (newPath: string) =>
    (pathDs.current[pathIdx.current] += newPath)

  // NOTE: We're dynamically assigning the path prop to avoid rerendering
  // on every gesture / drawing of the line
  const setPathProps = (path: string) => {
    if (pathEls[pathIdx.current]) {
      pathEls[pathIdx.current].setNativeProps({
        d: path,
      })
    }
  }

  const handleDrawStart = (e: GestureResponderEvent): void => {
    const { curX, curY } = extractCoords(e)
    action(curX, curY)
    modifyPath(`M${curX},${curY}`)

    coords.current = {
      curX,
      curY,
      prevX: curX,
      prevY: curY,
    }
    forceUpdate()
  }

  const handleDraw = (e: GestureResponderEvent): void => {
    const { prevX, prevY } = coords.current
    const { curX, curY } = extractCoords(e)
    const newCoords = { curX, curY, prevX, prevY }
    action(curX, curY)

    if (shouldComputeLine(newCoords, MIN_DISTANCE_SQ)) {
      // NOTE(@clauxx): take each line in the path and compare
      // it to the last one
      findIntersections(linesPts.current, newCoords).map((intersection) => {
        setCircles([...circles, [intersection.x, intersection.y]])
        setCirclesN(circles.length)
      })

      // NOTE: then add the new line and maintain the line list
      if (linesPts.current.length >= MAX_LINE_PTS) {
        const isMaxPath =
          pathIdx.current < pathDs.current.length - 1 &&
          (pathIdx.current == 0 ||
            linesPtsIdx.current >= linesPts.current.length)
        if (isMaxPath) {
          pathIdx.current++
          modifyPath(`M${prevX},${prevY}L${curX},${curY}`)
          forceUpdate()
        }

        if (linesPtsIdx.current >= linesPts.current.length)
          linesPtsIdx.current = 0

        linesPts.current.splice(
          linesPtsIdx.current,
          4,
          prevX,
          prevY,
          curX,
          curY,
        )
        linesPtsIdx.current += 4
      } else {
        linesPts.current = [...linesPts.current, prevX, prevY, curX, curY]
      }

      coords.current = {
        curX,
        curY,
        prevX: curX,
        prevY: curY,
      }

      modifyPath(`L${curX},${curY}`)
      setPathProps(pathDs.current[pathIdx.current])
    } else {
      // if the line segment was too short, we don't commit it to state but we
      // draw it anyway
      coords.current = {
        ...coords.current,
        curX,
        curY,
      }

      setPathProps(pathDs.current[pathIdx.current] + `L${curX},${curY}`)
    }
  }

  return { handleDrawStart, handleDraw, pathDs, pathEls, circles }
}
