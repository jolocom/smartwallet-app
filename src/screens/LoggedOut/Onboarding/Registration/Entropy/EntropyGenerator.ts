import * as sjcl from 'sjcl'

export interface EntropyGeneratorInterface {
  addFromDelta: (d: number) => void
  getProgress: () => number
  generateRandomString: (wordCount: number) => string
}

export class EntropyGenerator implements EntropyGeneratorInterface {
  private generator = new sjcl.prng(10)

  addFromDelta(d: number): void {
    this.generator.addEntropy(d, 1, 'user')
  }

  getProgress(): number {
    return this.generator.getProgress()
  }

  /**
   * Once the generator is ready, it can be used to generate a random buffer
   * @param wordCount - Amount of random words to generate (1 word - 4 bytes)
   * @returns Hex string encoding wordCount * 4 bytes random bytes
   */

  generateRandomString(wordCount: number): string {
    const intArray = new Int32Array(this.generator.randomWords(wordCount))
    const buf = Buffer.from(intArray.buffer)
    return buf.toString('hex')
  }
}
