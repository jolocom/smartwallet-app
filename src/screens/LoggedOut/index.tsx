import React from 'react'
import { createStackNavigator } from '@react-navigation/stack'

import { ScreenNames } from '~/types/screens'

import Walkthrough from './Walkthrough'
import Entropy from './Entropy'
import Recovery from './Recovery'
import SeedPhrase from './SeedPhrase'
import SeedPhraseRepeat from './SeedPhraseRepeat'

const Stack = createStackNavigator()

const LoggedOut: React.FC = () => {
  return (
    <Stack.Navigator
      headerMode="none"
      initialRouteName={ScreenNames.Walkthrough}
    >
      <Stack.Screen name={ScreenNames.Walkthrough} component={Walkthrough} />
      <Stack.Screen name={ScreenNames.Entropy} component={Entropy} />
      <Stack.Screen name={ScreenNames.Recovery} component={Recovery} />
      <Stack.Screen name={ScreenNames.SeedPhrase} component={SeedPhrase} />
      <Stack.Screen
        name={ScreenNames.SeedPhraseRepeat}
        component={SeedPhraseRepeat}
      />
    </Stack.Navigator>
  )
}

export default LoggedOut
