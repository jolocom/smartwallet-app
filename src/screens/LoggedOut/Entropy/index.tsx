import React, { useRef, useState, useEffect, useCallback } from 'react'
import { View, StyleSheet, Platform } from 'react-native'

import ScreenContainer from '~/components/ScreenContainer'

import { useReplaceWith } from '~/hooks/navigation'
import { ScreenNames } from '~/types/screens'
import { generateSecureRandomBytes } from '~/utils/generateBytes'

import { EntropyIntro } from './EntropyIntro'
import { EntropyGenerator } from './EntropyGenerator'
import { EntropyCanvas } from './EntropyCanvas'
import { Colors } from '~/utils/colors'
import JoloText, { JoloTextKind } from '~/components/JoloText'
import { JoloTextSizes } from '~/utils/fonts'
import { useGenerateSeed } from '~/hooks/sdk'
import { useLoader } from '~/hooks/useLoader'

//NOTE: Determines the duration of entropy collection
const ENOUGH_ENTROPY_PROGRESS = Platform.select({
  ios: 0.6,
  android: 0.3,
  default: 0.3,
})

const Entropy: React.FC = () => {
  const redirectToSeedPhrase = useReplaceWith(ScreenNames.SeedPhrase)
  const refreshEntropy = useReplaceWith(ScreenNames.Entropy)
  //const createIdentity = useIdentityCreate()
  const generateSeed = useGenerateSeed()
  const loader = useLoader()

  //NOTE: not using the user generated entropy
  const submitEntropy = async (entropy: string) => {
    const success = await loader(generateSeed, { showSuccess: false })

    if (success) {
      return redirectToSeedPhrase()
    }
    return refreshEntropy()
  }

  const { entropyProgress, addPoint } = useEntropyProgress(submitEntropy)

  return (
    <ScreenContainer>
      {!!entropyProgress ? (
        <View style={styles.percentage}>
          <JoloText
            kind={JoloTextKind.title}
            size={JoloTextSizes.middle}
            color={Colors.white85}
          >{`${Math.trunc(entropyProgress * 100)} %`}</JoloText>
        </View>
      ) : (
        <EntropyIntro />
      )}
      <EntropyCanvas disabled={entropyProgress === 1} addPoint={addPoint} />
    </ScreenContainer>
  )
}

export const useEntropyProgress = (submit: (entropy: string) => void) => {
  const [entropyProgress, setProgress] = useState(0)

  let entropyGenerator = useRef(new EntropyGenerator())

  useEffect(() => {
    if (entropyProgress === 1) {
      ;(async () => {
        await supplementEntropyProgress()
        submit(entropyGenerator.current.generateRandomString(4))
      })()
    }
  }, [entropyProgress])

  const supplementEntropyProgress = async () => {
    while (entropyGenerator.current.getProgress() < 1) {
      const moreEntropy = await generateSecureRandomBytes(512)
      // NOTE do not use moreEntropy.forEach, Buffer API is inconsistent, it
      // doesn't work in some envirtonments
      for (let i = 0; i < moreEntropy.length; i++) {
        entropyGenerator.current.addFromDelta(moreEntropy[i])
      }
    }
  }

  const updateProgress = () => {
    const progress =
      entropyGenerator.current.getProgress() / ENOUGH_ENTROPY_PROGRESS
    setProgress(progress >= 1 ? 1 : progress)
  }

  const addPoint = useCallback((x: number, y: number) => {
    entropyGenerator.current.addFromDelta(x)
    entropyGenerator.current.addFromDelta(y)

    updateProgress()
  }, [])

  return { entropyProgress, addPoint, entropyGenerator }
}

const styles = StyleSheet.create({
  percentage: {
    position: 'absolute',
    bottom: '10%',
    alignItems: 'center',
    paddingHorizontal: '5%',
  },
})

export default Entropy
