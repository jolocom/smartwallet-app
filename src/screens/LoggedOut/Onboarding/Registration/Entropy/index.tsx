import React, { useEffect } from 'react'

import { useReplaceWith } from '~/hooks/navigation'
import { ScreenNames } from '~/types/screens'

import { useGenerateSeed } from '~/hooks/sdk'
import { useLoader } from '~/hooks/loader'
import useTranslation from '~/hooks/useTranslation'

const Entropy: React.FC = () => {
  const replaceWith = useReplaceWith()
  const generateSeed = useGenerateSeed()
  const loader = useLoader()
  const { t } = useTranslation()

  const submitEntropy = async () => {
    const handleDone = (error: any) => {
      if (!error) {
        return replaceWith(ScreenNames.SeedPhraseWrite)
      }
      return replaceWith(ScreenNames.Entropy)
    }
    await loader(
      generateSeed,
      { showSuccess: false, loading: t('Entropy.loader') },
      handleDone,
    )
  }

  useEffect(() => {
    submitEntropy()
  }, [])

  return null
}

export default Entropy
