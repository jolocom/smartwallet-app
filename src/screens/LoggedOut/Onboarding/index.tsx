import React from 'react'
import { createStackNavigator } from '@react-navigation/stack'

import { ScreenNames } from '~/types/screens'
import Recovery from '~/screens/Modals/Recovery'
import TermsConsent from '~/screens/Modals/TermsConsent'
import { useSelector } from 'react-redux'
import { shouldShowTermsConsent } from '~/modules/account/selectors'

const Stack = createStackNavigator()

const Onboarding: React.FC = () => {
  const shouldShowTerms = useSelector(shouldShowTermsConsent)

  return (
    <Stack.Navigator headerMode="none">
      {shouldShowTerms ? (
        <Stack.Screen
          name={ScreenNames.LoggedOutTermsConsent}
          component={TermsConsent}
        />
      ) : (
        <Stack.Screen
          name={ScreenNames.IdentityRecovery}
          component={Recovery}
        />
      )}
    </Stack.Navigator>
  )
}

export default Onboarding
