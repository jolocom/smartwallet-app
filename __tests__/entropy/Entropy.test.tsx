import { useEntropyProgress } from '~/screens/LoggedOut/Entropy'
import { renderHook, act } from '@testing-library/react-hooks'

describe('Entropy', () => {
  it('should match snapshot after adding a point to the canvas', () => {
    const mockSubmit = jest.fn()
    const { result } = renderHook(() => useEntropyProgress(mockSubmit))
    const { addPoint, entropyGenerator } = result.current
    jest.spyOn(entropyGenerator.current, 'addFromDelta')

    act(() => {
      addPoint(100, 300)
    })

    expect(entropyGenerator.current.addFromDelta).toBeCalledTimes(2)
    expect(result.current.entropyProgress).toMatchSnapshot()
  })

  it('should successfully increment progress after adding a line', () => {
    const mockSubmit = jest.fn()
    const { result } = renderHook(() => useEntropyProgress(mockSubmit))
    const { addPoint, entropyGenerator } = result.current
    jest.spyOn(entropyGenerator.current, 'addFromDelta')

    act(() => {
      addPoint(100, 300)
      addPoint(200, 500)
      addPoint(400, 200)
    })

    expect(entropyGenerator.current.addFromDelta).toBeCalledTimes(6)
    expect(result.current.entropyProgress).toMatchSnapshot()
  })

  it('should successfully generate the entropy', () => {
    const mockSubmit = jest.fn()
    const { result } = renderHook(() => useEntropyProgress(mockSubmit))
    const { addPoint } = result.current

    const randomCoord = () => Math.floor(Math.random() * Math.floor(500))
    act(() => {
      for (let i = 0; i < 200; i++) addPoint(randomCoord(), randomCoord())
    })

    expect(result.current.entropyProgress).toMatchSnapshot()
    expect(mockSubmit).toBeCalled
  })
})
