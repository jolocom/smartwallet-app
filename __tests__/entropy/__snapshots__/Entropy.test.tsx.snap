// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entropy should match snapshot after adding a point to the canvas 1`] = `0.006510416666666667`;

exports[`Entropy should successfully generate the entropy 1`] = `1`;

exports[`Entropy should successfully increment progress after adding a line 1`] = `0.01953125`;
